lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,64 +1175,193 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
         EntriesString += MS->getString();
         EntriesString += '\0';
       }
     }
     auto EntriesConstant = llvm::ConstantDataArray::getString(
-        getLLVMContext(), EntriesString, /*AddNull=*/false);
+        IGM.getLLVMContext(), EntriesString, /*AddNull=*/false);
     // Mark the swift1_autolink_entries section with the SHF_EXCLUDE attribute
     // to get the linker to drop it in the final linked binary.
-    // LLVM doesn't provide an interface to specify section attributs in the IR
-    // so we pass the attribute with inline assembly.
-    if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF)
-      Module.appendModuleInlineAsm(".section .swift1_autolink_entries,"
-                                   "\"0x80000000\"");
+    // LLVM doesn't provide an interface to specify section attributs in the
+    // IR so we pass the attribute with inline assembly.
+    if (IGM.TargetInfo.OutputObjectFormat == llvm::Triple::ELF)
+      IGM.Module.appendModuleInlineAsm(".section .swift1_autolink_entries,"
+                                       "\"0x80000000\"");
     auto var =
-        new llvm::GlobalVariable(*getModule(), EntriesConstant->getType(), true,
-                                 llvm::GlobalValue::PrivateLinkage,
+        new llvm::GlobalVariable(*IGM.getModule(), EntriesConstant->getType(),
+                                 true, llvm::GlobalValue::PrivateLinkage,
                                  EntriesConstant, "_swift1_autolink_entries");
     var->setSection(".swift1_autolink_entries");
-    var->setAlignment(llvm::MaybeAlign(getPointerAlignment().getValue()));
+    var->setAlignment(llvm::MaybeAlign(IGM.getPointerAlignment().getValue()));
+
+    disableAddressSanitizer(IGM, var);
+    IGM.addUsedGlobal(var);
+    return;
+  }
+  }
+  llvm_unreachable("Unhandled AutolinkKind in switch.");
+}
 
-    disableAddressSanitizer(*this, var);
-    addUsedGlobal(var);
+AutolinkKind AutolinkKind::create(const SwiftTargetInfo &TargetInfo,
+                                  llvm::Triple Triple,
+                                  IRGenLLVMLTOKind LLVMLTOKind) {
+  // When performing LTO, we always use lld that supports auto linking
+  // mechanism with ELF. So embed dependent libraries names in
+  // "llvm.dependent-libraries" instead of "llvm.linker.options".
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF &&
+      LLVMLTOKind != IRGenLLVMLTOKind::None) {
+    return AutolinkKind::LLVMDependentLibraries;
   }
 
+  if (doesTargetAutolinkUsingAutolinkExtract(TargetInfo, Triple)) {
+    return AutolinkKind::SwiftAutoLinkExtract;
+  }
+
+  return AutolinkKind::LLVMLinkerOptions;
+}
+
+void IRGenModule::emitAutolinkInfo() {
+  auto Autolink =
+      AutolinkKind::create(TargetInfo, Triple, IRGen.Opts.LLVMLTOKind);
+
+  StringRef AutolinkSectionName = Autolink.getSectionNameMetadata();
+
+  auto *Metadata = Module.getOrInsertNamedMetadata(AutolinkSectionName);
+  llvm::SetVector<llvm::MDNode *> Entries;
```

Comment:
I think this was a SmallVector before, no? Why not make a SmallSetVector? Or maybe my memory is wrong.

Reply:
Ah, yes. I restored to use SmallSetVector